ocaml/testsuite/tests/typing-layouts/allow-illegal-crossing.ml

```diff
@@ -0,0 +1,654 @@
+and B : sig type t : value mod portable uncontended end
+|}]
+
+type t : value mod portable = private { foo : string }
+[%%expect {|
+type t = private { foo : string; }
+|}]
+
+type a : value mod portable = { foo : string }
+type b : value mod portable = a = { foo : string }
+[%%expect {|
+type a = { foo : string; }
+type b = a = { foo : string; }
+|}]
+
+type a : value mod uncontended = private { foo : string }
+type b : value mod uncontended = a = private { foo : string }
+[%%expect {|
+type a = private { foo : string; }
+type b = a = private { foo : string; }
+|}]
+
+type t : value mod uncontended = private Foo of int | Bar
+[%%expect {|
+type t = private Foo of int | Bar
+|}]
+
+type a : value mod uncontended = Foo of int | Bar
+type b : value mod uncontended = a = Foo of int | Bar
+[%%expect {|
+type a = Foo of int | Bar
+type b = a = Foo of int | Bar
+|}]
+
+type a : value mod portable = private Foo of int | Bar
+type b : value mod portable = a = private Foo of int | Bar
+[%%expect {|
+type a = private Foo of int | Bar
+type b = a = private Foo of int | Bar
+|}]
+
+(********************************************)
+(* Test 2: illegal mode crossing propogates *)
+
+type a : value mod portable uncontended = private string
+type ('a : value mod portable uncontended) b
+type c = a b
+[%%expect {|
+type a = private string
+type ('a : value mod uncontended portable) b
+type c = a b
+|}]
+
+type t : value mod portable uncontended = private string
+let f : ('a : value mod portable uncontended). 'a -> 'a = fun x -> x
+let g (x : t) = f x
+[%%expect {|
+type t = private string
+val f : ('a : value mod uncontended portable). 'a -> 'a = <fun>
+val g : t -> t = <fun>
+|}]
+
+type t : value mod portable uncontended = { a : int; b : int }
+let x : _ as (_ : value mod portable uncontended) = { a = 5; b = 5 }
+[%%expect {|
+type t = { a : int; b : int; }
+val x : t = {a = 5; b = 5}
+|}]
+
+type ('a, 'b) t : value mod portable uncontended = { a : 'a; b : 'b }
+let x : _ as (_ : value mod portable uncontended) = { a = 5; b = 5 }
+[%%expect {|
+type ('a, 'b) t = { a : 'a; b : 'b; }
+val x : (int, int) t = {a = 5; b = 5}
+|}]
+
+type t : value mod portable uncontended = Foo of string | Bar of int
+let x : _ as (_ : value mod portable uncontended) = Foo "hello world"
+[%%expect {|
+type t = Foo of string | Bar of int
+val x : t = Foo "hello world"
+|}]
+
+(********************************************************)
+(* Test 3: types cannot cross other axes when annotated *)
+
+type a : value mod local
+type b : value mod global = private a
+[%%expect {|
+type a : value mod local
+Line 2, characters 0-37:
+2 | type b : value mod global = private a
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The layout of type a is value, because
+         of the definition of a at line 1, characters 0-24.
+       But the layout of type a must be a sublayout of value, because
+         of the definition of b at line 2, characters 0-37.
+|}]
+
+type a : value mod once
```

Comment:
`mod once` should be a no-op. Leave it out (here and elsewhere where `mod`ding by top).